File: javascript/packages/linter/src/rules/erb-no-javascript-tag-helper.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"
import { getTagLocalName, isERBOpenTagNode, HELPER_REGISTRY } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult, ParserOptions, HTMLElementNode } from "@herb-tools/core"

const JAVASCRIPT_TAG_ELEMENT_SOURCE = HELPER_REGISTRY["javascript_tag"].source

class ERBNoJavascriptTagHelperVisitor extends BaseRuleVisitor {
  visitHTMLElementNode(node: HTMLElementNode): void {
    if (this.isJavascriptTagHelper(node)) {
      this.addOffense(
        "Avoid `javascript_tag`. Use inline `<script>` tags instead.",
        node.open_tag!.location,
      )
    }

    super.visitHTMLElementNode(node)
  }

  private isJavascriptTagHelper(node: HTMLElementNode): node is HTMLElementNode {
    if (getTagLocalName(node) !== "script") return false
    if (!isERBOpenTagNode(node.open_tag)) return false

    return node.element_source === JAVASCRIPT_TAG_ELEMENT_SOURCE
  }
}

// TODO: `javascript_tag` is an Action View helper, so this rule belongs under the `actionview-` prefix. The rename waits on a rule-alias mechanism, since it breaks every `.herb.yml` and `herb:disable` comment naming it.
export class ERBNoJavascriptTagHelperRule extends ParserRule {
  static ruleName = "erb-no-javascript-tag-helper"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning",
      frameworks: ["actionview"],
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoJavascriptTagHelperVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/erb_no_javascript_tag_helper.rs
use crate::utils::tag_utils::get_tag_local_name;

use herb::nodes::HTMLElementNode;
use herb::union_types::ERBOpenTagNodeOrHTMLConditionalOpenTagNodeOrHTMLOpenTagNode;
use herb::Visitor;

rule_visitor!(ERBNoJavascriptTagHelperVisitor);
define_parser_rule!(
  ERBNoJavascriptTagHelperRule,
  "erb-no-javascript-tag-helper",
  Warning,
  ERBNoJavascriptTagHelperVisitor,
  parser_options: { action_view_helpers: true }
);

fn javascript_tag_element_source() -> &'static str {
  herb::action_view_helpers::find_by_name("javascript_tag")
    .map(|entry| entry.source)
    .unwrap_or("")
}

impl Visitor for ERBNoJavascriptTagHelperVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if get_tag_local_name(node).as_deref() == Some("script") {
      if let Some(ERBOpenTagNodeOrHTMLConditionalOpenTagNodeOrHTMLOpenTagNode::ERBOpenTagNode(open_tag)) = &node.open_tag {
        if node.element_source == javascript_tag_element_source() {
          self.add_offense("Avoid `javascript_tag`. Use inline `<script>` tags instead.", open_tag.location.clone());
        }
      }
    }

    self.walk_html_element_node(node);
  }
}
